src/main/java/jcifs/smb1/smb1/NtlmPasswordAuthentication.java

```diff
@@ -0,0 +1,636 @@
+     * @param clientChallenge The client challenge (nonce). 
+     */ 
+    public static byte[] getLMv2Response(String domain, String user,
+            String password, byte[] challenge, byte[] clientChallenge) {
+        try {
+            byte[] hash = new byte[16];
+            byte[] response = new byte[24];
+// The next 2-1/2 lines of this should be placed with nTOWFv1 in place of password
+            MD4 md4 = new MD4();
+            md4.update(password.getBytes(SmbConstants.UNI_ENCODING));
+            HMACT64 hmac = new HMACT64(md4.digest());
+            hmac.update(user.toUpperCase().getBytes(SmbConstants.UNI_ENCODING));
+            hmac.update(domain.toUpperCase().getBytes(SmbConstants.UNI_ENCODING));
+            hmac = new HMACT64(hmac.digest());
+            hmac.update(challenge);
+            hmac.update(clientChallenge);
+            hmac.digest(response, 0, 16);
+            System.arraycopy(clientChallenge, 0, response, 16, 8);
+            return response;
+        } catch (Exception ex) {
+            if( log.level > 0 )
+                ex.printStackTrace( log );
+            return null;
+        }
+    }
+    public static byte[] getNTLM2Response(byte[] nTOWFv1,
+                    byte[] serverChallenge,
+                    byte[] clientChallenge)
+    {
+        byte[] sessionHash = new byte[8];
+
+        try {
+            MessageDigest md5;
+            md5 = MessageDigest.getInstance("MD5");
+            md5.update(serverChallenge);
+            md5.update(clientChallenge, 0, 8);
+            System.arraycopy(md5.digest(), 0, sessionHash, 0, 8);
+        } catch (GeneralSecurityException gse) {
+            if (log.level > 0)
+                gse.printStackTrace(log);
+            throw new RuntimeException("MD5", gse);
+        }
+
+        byte[] key = new byte[21];
+        System.arraycopy(nTOWFv1, 0, key, 0, 16);
+        byte[] ntResponse = new byte[24];
+        E(key, sessionHash, ntResponse);
+
+        return ntResponse;
+    }
+    public static byte[] nTOWFv1(String password)
+    {
+        if (password == null)
+            throw new RuntimeException("Password parameter is required");
+        try {
+            MD4 md4 = new MD4();
+            md4.update(password.getBytes(SmbConstants.UNI_ENCODING));
+            return md4.digest();
+        } catch (UnsupportedEncodingException uee) {
+            throw new RuntimeException(uee.getMessage());
+        }
+    }
+    public static byte[] nTOWFv2(String domain, String username, String password)
+    {
+        try {
+            MD4 md4 = new MD4();
+            md4.update(password.getBytes(SmbConstants.UNI_ENCODING));
+            HMACT64 hmac = new HMACT64(md4.digest());
+            hmac.update(username.toUpperCase().getBytes(SmbConstants.UNI_ENCODING));
+            hmac.update(domain.getBytes(SmbConstants.UNI_ENCODING));
+            return hmac.digest();
+        } catch (UnsupportedEncodingException uee) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)